Add higher level export constraints api

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96012
* __->__ #96338
* #96678

This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Add higher level export constraints api

[ghstack-poisoned]

gmagogsfm commented on torch/_export/__init__.py:
```diff
@@ -203,3 +203,40 @@ def do_not_use_experimental_export(f: Callable, args: Tuple, training=False):
     graph_module, _, out_spec = _aot_capture(graph_module, flat_args)
     return ExportedProgram(fw_module=graph_module, example_inputs=original_flat_args, in_spec=in_spec, out_spec=out_spec)
+
+Constraint = Tuple[torch.Tensor, int]
```
Some docstring would be helpful to explain what the `int` is.

gmagogsfm commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1534,7 +1534,9 @@ def _verify(expr, potential_expr):
-                        raise RuntimeError(f"Attempting to constrain dim {i} for {source}, which violates user's mark_dynamic")
+                        raise RuntimeError(f"Attempting to constrain dimension "
+                                           f"{source.name()}.size()[{i}] to {int(ss)}, "
+                                           "which violates user's mark_dynamic")
```
```suggestion
                                           "which violates user's constraints")
```

gmagogsfm commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1514,7 +1514,7 @@ def _verify(expr, potential_expr):
                 for src in srcs:
                     if src in dynamic_sources:
-                        raise RuntimeError(f"Attempting to introduce a guard {potential_expr} that violates user's mark_dynamic")
+                        raise RuntimeError(f"Attempting to introduce a guard {potential_expr} that violates user's constraint")
```
```suggestion
                        raise RuntimeError(f"Attempting to introduce a guard {potential_expr} that violates user's constraints")
```

gmagogsfm commented on test/dynamo/test_export.py:
```diff
@@ -2190,6 +2183,59 @@ def forward(self, *args):
+
+        def my_dyn_fn(x):
+            return x.cos()
+
```
Nit:
Construct dynamic_dim before `export` call and asserts that dynamic_dim doesn't add state to `y` either.

voznesenskym replied:
good call

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -77,7 +78,7 @@ def _profiler_end(record):
 
 # TODO can we enable by default? (check perf CI) otherwise, guard behind config
-enable_cache_lookup_profiler(config.profile_cache_lookup)
+# enable_cache_lookup_profiler(config.profile_cache_lookup)
```
yes

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +697,174 @@ def result_capturing_wrapper(*graph_inputs):
+        new_graph.recompile()
+    finally:
+        for tensor in tensor_to_indices.keys():
+            torch._dynamo.clear_dynamic_dims(tensor)
```
Maybe, but I do like having that safety... and assertion in clearing feels serious enough that you would expect state violations after....

voznesenskym committed:
Update on "Add higher level export constraints api"




cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

gchanan commented:
what is this PR trying to do?  I don't see any description.

I am not convinced that an API at the export level makes sense.  How do you know you don't need to annotate the model outside of the export context?

voznesenskym commented:
> what is this PR trying to do? I don't see any description.
> 
> I am not convinced that an API at the export level makes sense. How do you know you don't need to annotate the model outside of the export context?

Hey! This API is the result of a few hours design sesh with @ezyang @avikchaudhuri @tugsbayasgalan and @gmagogsfm.

> How do you know you don't need to annotate the model outside of the export context?

This is an export only api! The eager API still exists, and is still in `torch._dynamo.mark_dynamic`, which this just uses under the hood as a higher level API.

gchanan commented:
> This is an export only api! The eager API still exists, and is still in torch._dynamo.mark_dynamic, which this just uses under the hood as a higher level API.

Right.  How do you know you don't _also_ need to annotate inside the model?  If in practice you need to, what's the point of having a separate API?

ezyang commented on torch/_export/__init__.py:
```diff
@@ -203,3 +203,41 @@ def do_not_use_experimental_export(f: Callable, args: Tuple, training=False):
+#     ]
+# )
+def dynamic_dim(t: torch.Tensor, index: int):
+    return (t, index)
```
You sure you want to represent this as just a tuple? Seems better to chuck it in a dataclass or something.

ezyang replied:
Just do it now, you already know you are going to need the generality.

voznesenskym replied:
sg

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
> > This is an export only api! The eager API still exists, and is still in torch._dynamo.mark_dynamic, which this just uses under the hood as a higher level API.
> 
> Right. How do you know you don't _also_ need to annotate inside the model? If in practice you need to, what's the point of having a separate API?

We never annotate inside the model today

ezyang commented on torch/_dynamo/__init__.py:
```diff
@@ -167,3 +167,14 @@ def mark_dynamic(t, index):
+@forbid_in_graph
+def clear_dynamic_dims(t):
+    assert has_dynamic_dims(t)
+    delattr(t, "_dynamo_dynamic_indices")
```
These aren't user facing right? Let's not put these in the top-level `torch._dynamo` namespace. Put them somewhere else.

voznesenskym replied:
They are user facing in the lower level API, not the higher one. It seems like a good complement to have alongside mark_dynamic, right?

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -608,6 +609,10 @@ def export(
         tracing_mode (str): Specifies the tracing mode. Must be set to "real" if decomposition_table is not specified.
         If decomposition_table is specified, the options are "symbolic" or "fake". Default is "real".
 
+        constraints List[Constraint]: A list of user specified constraints, see Note - [On Export Dynamic Dimension UX]
```
You can't reference Notes from docblocks, they won't actually show up in the generated documentation. Please properly format, or, probably easier, just put the proper discussion here in the docblock.

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+        for constraint in constraints:
+            tensor = constraint[0]
+            index = constraint[1]
+            # TODO(voz): Is tensor defaultdict friendly?
```
Use WeakIdKeyDictionary from torch.utils.weak, which does proper id semantics even though Tensor defines value equality for ==.

ezyang replied:
Or just map from id to int, this dict is very short lived.

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -608,6 +609,10 @@ def export(
 
+        constraints List[Constraint]: A list of user specified constraints, see Note - [On Export Dynamic Dimension UX]
+        for more details. Constraints must be unique per tensor+dim pair, must not be called alongside mark_dynamic
+        for the same tensor, and must only specify tensors that are part of the arg inputs.
```
Also, please learn how to properly format docblocks. Read the ReST manual or something, or go look at some of our existing docblocks.

ezyang commented on torch/_export/__init__.py:
```diff
@@ -203,3 +203,41 @@ def do_not_use_experimental_export(f: Callable, args: Tuple, training=False):
+# After a lot of discussion, we have settled on a dynamic marking API
+# for export that meets the following constraints:
+# 1) Stateless
+# 2) Safe for numerous .export calls
```
it is not obvious from context what "numerous .export calls" means

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+            if tensor not in tensor_to_indices:
+                assert not torch._dynamo.has_dynamic_dims(
+                    tensor
+                ), "Illegal to export tensor already marked dynamic"
```
This makes the implementation here simpler, but it doesn't make any sense from a UX perspective.  Preexisting markings should be ignored, and left undisturbed.

ezyang replied:
Don't use assert for user error. Raise a regular RuntimeError

voznesenskym replied:
I don't understand the use case for randomly calling export in the middle of something, where there are already marked dims? It just seems like an accidental / error case. That being said, If you feel strongly about this, I am happy to rewrite this to support restoring state.

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+                    tensor
+                ), "Illegal to export tensor already marked dynamic"
+                tensor_to_indices[tensor] = set()
+            assert index not in tensor_to_indices[tensor], f"Duplicate index {index}"
```
Don't use assert for user error. Raise a regular RuntimeError

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+                tensor_to_indices[tensor] = set()
+            assert index not in tensor_to_indices[tensor], f"Duplicate index {index}"
+            tensor_to_indices[tensor].add(index)
+            torch._dynamo.mark_dynamic(tensor, index)
```
You're going to have to rewrite all this code when you add range constraints...

voznesenskym replied:
Yes, but we don't have the mechanism for those yet, so I am fine with that for now.

gchanan commented:
> We never annotate inside the model today, in fact, that is illegal in the current state of things.

That's not quite what I was asking -- my question was unclear.  Let me try to rephrase:

- What we currently have is a situation where we specify constraints on the inputs
- "We never annotate inside the model today, in fact, that is illegal in the current state of things."

How do you know (in practice?) this will be usable for export?  I.e. that there is dynamism inside the model that isn't captured (or propagated? -- unclear to me how this works) by input tensor constraints.

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+        for arg in flat_args:
+            if isinstance(arg, torch.Tensor) and arg in seen_tensors:
+                seen_tensors.remove(arg)
+        assert len(seen_tensors) == 0, "User specified dynamic dim for unknown tensors"
```
nit: this condition can be greatly simplified as `tensor_to_indices.keys() - set(flat_args)`

voznesenskym replied:
Not if we have non tensor args, but we can simplify it a bit

voznesenskym commented:
> > We never annotate inside the model today, in fact, that is illegal in the current state of things.
> 
> That's not quite what I was asking -- my question was unclear. Let me try to rephrase:
> 
> * What we currently have is a situation where we specify constraints on the inputs
> * "We never annotate inside the model today, in fact, that is illegal in the current state of things."
> 
> How do you know (in practice?) this will be usable for export? I.e. that there is dynamism inside the model that isn't captured (or propagated? -- unclear to me how this works) by input tensor constraints.

Sorry, I am being a little dumb here. Are you asking what the difference between this, and input range constraints is?

tugsbayasgalan commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+        ).transform()
+
+        # Make dynamo graph to have same input/output spec as user code
+        def argument_names(f: Callable[..., Any], *args, **kwargs) -> List[str]:
```
NIT: i feel like this try block is too big now. Could we make these kind of things into separate utility methods so that the code is little shorter?

voznesenskym replied:
Not in this pr

tugsbayasgalan commented on torch/_dynamo/eval_frame.py:
```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
-            self.old_args_gen = (
-                self.new_args[i] for i in matched_input_elements_positions
-            )
+        class ChangeInputOutputSignature(torch.fx.interpreter.Transformer):
```
NIT: Can we move this out of export function?

voznesenskym replied:
Not in this pr

tugsbayasgalan commented:
> > We never annotate inside the model today, in fact, that is illegal in the current state of things.
> 
> That's not quite what I was asking -- my question was unclear. Let me try to rephrase:
> 
> * What we currently have is a situation where we specify constraints on the inputs
> * "We never annotate inside the model today, in fact, that is illegal in the current state of things."
> 
> How do you know (in practice?) this will be usable for export? I.e. that there is dynamism inside the model that isn't captured (or propagated? -- unclear to me how this works) by input tensor constraints.

In export case, we only really encounter very dimensions to be dynamic. If the user doesn't know about specific dynamism in the model code, i think the guards produced from torchdynamo can be used as extra constraint which the user can use to export again. cc: @avikchaudhuri

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add higher level export constraints api"


This API was the outcome a design session on a higher level export API.

A few notes from that discussion:

Design principles:
Suo: Stateless API (no mutable state for configuring the tensors)
Voz: The burden for diverging eager and export needs to be higher
Suo: Least surprise, if I trace with a different of constraints, it shouldn’t randomly fail if it failed on another trace

Constraints:
NON-GOAL: Handle people export multiple times
Testing commonly exports multiple times
Avoid use of context manager: it’s noise for what the user wants to do

Rejected proposals:
No intersecting constraints if you mark dynamic multiple times

We settled on a stateless API that passes a structure downstream to export. Export then, under the hood, unpacks the structure and under certain safety checks, installs the dynamic dims, runs export, and then cleans up after itself.

This api lives inside torch._export:

`def dynamic_dim(t: torch.Tensor, index: int)`

And is used as such:
```
y = torch.randn([3, 3, 3])
def my_dyn_fn(x):
    if x.shape[0] == 3:
        return x.sin()
    return x.cos()

graph, guards = torch._dynamo.export(my_dyn_fn, y, constraints=[dynamic_dim(y, 0)])
```

This now means that export has a different API from eager, in that eager is still advised to use `torch._dynamo.mark_dynamic`


cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
https://github.com/pytorch/pytorch/pull/96699
